### ISSUE TITLE
17.03 channel asks user to install unavailable and outdated version of oraclejdk8

### ISSUE THREAD

**Issue by mkocha2** [NixOS/nixpkgs#28632]
## Issue description

```Unfortunately, we cannot download file jdk-8u141-linux-x64.tar.gz automatically.
Please go to http://www.oracle.com/technetwork/java/javase/downloads/jdk8-downloads-2133151.html to download it yourself, and add it to the Nix store
using either
  nix-store --add-fixed sha256 jdk-8u141-linux-x64.tar.gz
or
  nix-prefetch-url --type sha256 file:///path/to/jdk-8u141-linux-x64.tar.gz
```

### Steps to reproduce
`nixos-rebuild switch -I nixpkgs=http://nixos.org/channels/nixos-17.03/nixexprs.tar.xz` with configuration.nix containing `oraclejdk8`.

## Technical details

* System: (NixOS: `nixos-version`, Ubuntu/Fedora: `lsb_release -a`, ...)
* Nix version: 1.11.8
* Nixpkgs version: (run `nix-instantiate --eval '<nixpkgs>' -A lib.nixpkgsVersion`)
* Sandboxing enabled: false

**Comment by moaxcp** [comment]
There is a java archive page for older releases. http://www.oracle.com/technetwork/java/javase/downloads/java-archive-javase8-2177648.html

**Comment by mkocha2** [comment]
That requires an account.

I'd also prefer not to install software which is not recommended by a vendor anymore. I expect that stable NixOS upgrades continue to work when vendors change some URLs from time to time.

**Comment by vcunat** [comment]
The purpose of stable channels is to receive only security updates and important bugfixes.  I don't know if this is the case; it's certainly not a simple "URL change" but an update.  We don't even have `oraclejdk8.meta.maintainers`...

As for Oracle requiring accounts etc., nixpkgs can't help that one at all.

**Comment by moaxcp** [comment]
I'm not sure how it works. According to [this](https://www.java.com/en/download/faq/release_dates.xml) update 144 is a limited update where update 141 is a critical update. I think the only way to follow the vendor's advice is to create a packageOverride for oraclejdk8.

**Comment by mkocha2** [comment]
Why does the following still try to download the 141 version?

```
    nixpkgs.config.packageOverrides = pkgs: {
      oraclejdk8 = pkgs.oraclejdk8.overrideAttrs (oldAttrs: {
	  productVersion = "8";
	  patchVersion = "144";
	  downloadUrl = http://www.oracle.com/technetwork/java/javase/downloads/jdk8-downloads-2133151.html;
	  sha256_i686 = "1i5pginc65xl5vxzwid21ykakmfkqn59v3g01vpr94v28w30jk32";
	  sha256_x86_64 = "1r5axvr8dg2qmr4zjanj73sk9x50m7p0w3vddz8c6ckgav7438z8";
	  sha256_armv7l = "10r3nyssx8piyjaspravwgj2bnq4537041pn0lz4fk5b3473kgfb";
	  jceName = "jce_policy-8.zip";
	  jceDownloadUrl = http://www.oracle.com/technetwork/java/javase/downloads/jce8-download-2133166.html;
	  sha256JCE = "0n8b6b8qmwb14lllk2lk1q1ahd3za9fnjigz5xn65mpg48whl0pk";
	  });
      };
  };

environment.systemPackages = with pkgs; [ oraclejdk8 ];
```

**Comment by vcunat** [comment]
These names are not (direct) attributes to the resulting `mkDerivation` call.

**Comment by mkocha2** [comment]
How can I see what the direct attributes are?

**Comment by vcunat** [comment]
The `mkDerivation` function is passed an attribute set/map, so those attributes.  I'm not sure what else is there to explain.

**Comment by mkocha2** [comment]
How about an example, since it's not clear at all what you mean?

**Comment by vyp** [comment]
@mkocha2 So I believe @vcunat is talking about [this](https://github.com/NixOS/nixpkgs/blob/5543e992f1d66aa4fb4e7ec19597bbe8b8ac18a8/pkgs/development/compilers/oraclejdk/jdk-linux-base.nix#L79) `mkDerivation` call. So if you look at the top of the file you can see these attributes being passed in, and you'll have to see how `mkDerivation` uses these passed in variables to override the correct attributes properly. For example I think you'll have to override the `name` attribute like so:

``` nix
{
  nixpkgs.config.packageOverrides = pkgs: {
    oraclejdk8 = let
      productVersion = "8";
      patchVersion = "144";
      downloadUrl = http://www.oracle.com/technetwork/java/javase/downloads/jdk8-downloads-2133151.html;
      sha256_i686 = "1i5pginc65xl5vxzwid21ykakmfkqn59v3g01vpr94v28w30jk32";
      sha256_x86_64 = "1r5axvr8dg2qmr4zjanj73sk9x50m7p0w3vddz8c6ckgav7438z8";
      sha256_armv7l = "10r3nyssx8piyjaspravwgj2bnq4537041pn0lz4fk5b3473kgfb";
      jceName = "jce_policy-8.zip";
      jceDownloadUrl = http://www.oracle.com/technetwork/java/javase/downloads/jce8-download-2133166.html;
      sha256JCE = "0n8b6b8qmwb14lllk2lk1q1ahd3za9fnjigz5xn65mpg48whl0pk";
    in pkgs.oraclejdk8.overrideAttrs (oldAttrs: {
      name = "oraclejdk-${productVersion}u${patchVersion}";
    });
  };
}
```

(And remember you can access the `mkDerivation` attributes from that file with `oldAttrs`.) I haven't tested this though! 😕 But I hope this helps somewhat.

**Comment by grahamc** [comment]
Fixed in 17.09 according to #29675.
